data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,452 @@
+library](https://github.com/ocsigen/tyxml), which uses the type system to ensure
+that the resulting HTML value satisfies most of the W3C standards. However,
+writing `Tyxml` values can be cumbersome compared to writing html syntax.
+Transforming HTML code into OCaml values at compile time allows to keep both the
+type-checker on the produced value and the familiarity of writing HTML code.
+
+Other rewriters such as `ppx_expect` show that being able to enrich the syntax
+via PPX rewriters is very useful, even outside of the specificity of OCaml.
+
+## The need for controlling the PPX ecosystem: Ppxlib
+
+Although PPXes are great for generating code at compile time, they raise a few
+questions, especially in the presence of multiple PPX rewriters:
+- What is the semantics of the composition of multiple PPXes? The order might
+  matter.
+- How can I trust that some part of my code are not modified, when using
+  third-party PPXes?
+- How can I keep short compilation time when using many rewriters?
+- How can I write a PPX easily if I have to deal with parsing or demarshalling
+  an AST, and deal with such a [long and complex
+  type](https://v2.ocaml.org/api/compilerlibref/Parsetree.html) as in parsetree?
+- How can I solve the problem that the OCaml syntax and the parsetree types
+  change with OCaml version?
+
+The good answer to these questions is a platform tool to write PPXes:
+[ppxlib](https://github.com/ocaml-ppx/ppxlib/). It takes care of easing the
+process of writing a PPX, ensure that composition is not too much of a problem,
+and factors as much as possible the work in case of multiple rewriters applied
+to a file.
+
+The idea of ppxlib is that PPX authors can concentrate on their own part, the
+rewriting logic. Then, they can register their transformation, and ppxlib will
+be responsible for the rest of the work that all PPXes have to do: getting the
+parsetree, giving it back to the compiler, creating an executable with a good
+CLI. Multiple transformations can be registered to generate a single binary.
+
+### One PPX for multiple OCaml version
+
+One of the challenge in writing a PPX is the types of the `Parsetree` module,
+and the OCaml syntax can change on version bump. To keep a PPX compatible with a
+new version, it would have to update the transformation from the old
+syntax or parsetree type, to the new syntax or parsetree type. But, by doing so
+it would lose compatibility with the old OCaml version.
+
+Ppxlib deals with this issue by converting parsetree types to and from the
+latest version. A PPX author then only need to provide a transformation for the
+latest version of OCaml and get a PPX that works on any version of OCaml. For
+instance, when applying a PPX written for OCaml `5.0`, during a compilation with
+ocaml `4.08`, ppxlib would get the `4.08` parsetree, transform it into a `5.00`
+parsetree, use the registered PPX to transform the parsetree, and convert it
+back to a `4.08` parsetree to continue the compilation.
```

Comment:
```suggestion
### One PPX for Multiple OCaml Versions

One of the challenges in writing a PPX is the types of the Parsetree module because 
the OCaml syntax can change on version bump. To keep a PPX compatible with a
new version, it would have to update the transformation from the old
syntax or Parsetree type to the new syntax or Parsetree type. But, by doing so,
it would lose compatibility with the old OCaml version.

`ppxlib` deals with this issue by converting Parsetree types to and from the
latest version. A PPX author then only needs to provide a transformation for 
OCaml's latest version and get a PPX that works on any version of OCaml. For
instance, when applying a PPX written for OCaml 5.0 during a compilation with
OCaml 4.08, `ppxlib` would get the 4.08 Parsetree, transform it into a 5.00
Parsetree, use the registered PPX to transform the Parsetree, and convert it
back to a 4.08 Parsetree to continue the compilation.
```